File: OJ/leetcode/cpp/78_subsets.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void dfs(int start, vector<int>& stack, vector<vector<int>>& ans, vector<int>& nums) {
    ans.push_back(stack);
    for (int i = start; i < nums.size(); ++i) {
        stack.push_back(nums[i]);
        dfs(i + 1, stack, ans, nums);
        stack.pop_back();
    }
}

vector<vector<int>> subsets(vector<int>& nums) {
    vector<vector<int>> ans;
    vector<int> stack;
    dfs(0, stack, ans, nums);
    return ans;
}

int main() {
    vector<int> nums = {1, 2, 3};
    vector<vector<int>> ans = subsets(nums);
    for (vector<int> vec : ans) {
        for (int i : vec) {
            cout << i << " ";
        }
        cout << "\n";
    }
}

File: competitive_programming/leetcode/python/78_subsets.py
def iterativeBfsSolution(nums):
  n = len(nums)
  tree = [[i] for i in range(0, n)]
  index_ans = []
  count = n
  while count > 0:
    tmp = []
    for arr in tree:
      for i in range(arr[-1] + 1, n):
        tmp.append(arr + [i])
      index_ans.append(arr)
    tree = tmp
    count -= 1
  ans = [[]]
  for arr in index_ans:
    ans.append([nums[i] for i in arr])
  return ans


def dfsSolution(nums):
  def dfs(depth, nums, start, stack, ans):
    ans.append(stack[:])
    if depth == len(nums):
      return

    for i in range(start, len(nums)):
      stack.append(nums[i])
      dfs(depth + 1, nums, i + 1, stack, ans)
      stack.pop()

  ans = []
  dfs(0, nums, 0, [], ans)
  return ans  
  
print(dfsSolution([1, 2, 3]))
  


